File: src/java/org/voegtle/weatherstation/server/logic/HealthProvider.java
package org.voegtle.weatherstation.server.logic;

import org.voegtle.weatherstation.server.persistence.Health;
import org.voegtle.weatherstation.server.persistence.HealthDTO;
import org.voegtle.weatherstation.server.persistence.LocationProperties;
import org.voegtle.weatherstation.server.persistence.PersistenceManager;
import org.voegtle.weatherstation.server.util.DateUtil;

import javax.cache.Cache;
import javax.cache.CacheException;
import javax.cache.CacheFactory;
import javax.cache.CacheManager;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

public class HealthProvider {
  protected static final Logger log = Logger.getLogger("HealthLogger");
  private static final String HEALTH = "health";
  private Cache cache;
  private PersistenceManager pm;

  private DateUtil dateUtil;

  private Health health;

  public HealthProvider(PersistenceManager pm, LocationProperties locationProperties) {
    this.pm = pm;
    this.dateUtil = locationProperties.getDateUtil();
    try {
      CacheFactory cacheFactory = CacheManager.getInstance().getCacheFactory();
      Map properties = new HashMap<>();
      cache = cacheFactory.createCache(properties);
    } catch (CacheException e) {
      log.severe("CentralServlet: Could not instantiate Cache");
    }
  }

  public HealthDTO get() {
    Date today = dateUtil.getToday();
    HealthDTO health = (HealthDTO)cache.get(HEALTH);
    if (health == null || isOutdated(health, today)) {
      health = pm.selectHealth(today).toDTO();
    }
    return health;
  }

  public void update(HealthDTO health) {
    cache.put(HEALTH, health);
    pm.makePersistant(health);
  }

  private boolean isOutdated(HealthDTO health, Date today) {
    return today.after(health.getDay());
  }
}

File: src/main/java/org/voegtle/weatherstation/server/logic/HealthProvider.kt
package org.voegtle.weatherstation.server.logic

import com.google.appengine.api.memcache.MemcacheService
import com.google.appengine.api.memcache.MemcacheServiceFactory
import org.voegtle.weatherstation.server.persistence.HealthDTO
import org.voegtle.weatherstation.server.persistence.PersistenceManager
import org.voegtle.weatherstation.server.persistence.entities.LocationProperties
import org.voegtle.weatherstation.server.util.DateUtil
import java.util.*

class HealthProvider(private val pm: PersistenceManager, locationProperties: LocationProperties) {
  private val HEALTH = "health"
  private val cache: MemcacheService

  private val dateUtil: DateUtil = locationProperties.dateUtil

  init {
    cache = MemcacheServiceFactory.getMemcacheService()
  }

  fun get(): HealthDTO {
    val today = dateUtil.today()
    val health: HealthDTO? = cache[HEALTH] as HealthDTO?
    if (health == null || isOutdated(health, today)) {
      return pm.selectHealth(today).toDTO()
    }
    return health
  }

  operator fun get(day: Date): HealthDTO = pm.selectHealth(day).toDTO()

  fun update(health: HealthDTO) {
    cache.put(HEALTH, health)
    pm.makePersistent(health)
  }

  private fun isOutdated(health: HealthDTO, today: Date): Boolean = today.after(health.day)

}
